File: python/rakka/profiler/_scenarios.py
"""Scenario actors + runners for the Python profiler."""

from __future__ import annotations

import time
from typing import Iterable, List, Optional

from rakka import Actor, ActorRef, ActorSystem, props

from ._probes import best_dispatcher_for, cpu_time_ns, peak_rss_bytes, rss_bytes
from ._report import Measurement, percentile


# --- actors (module-level so subinterpreters can import them) ---------------


class Sink(Actor):
    """Counts messages. ``ask('__count__')`` flushes and returns the total."""

    def __init__(self) -> None:
        self.n = 0

    async def handle(self, ctx, msg):
        if isinstance(msg, str) and msg == "__count__":
            return self.n
        self.n += 1
        return None


class Echo(Actor):
    async def handle(self, ctx, msg):
        return msg


class CpuWorker(Actor):
    """Each message runs a small hashing loop (matches the Rust scenario)."""

    async def handle(self, ctx, msg):
        h = (int(msg) if isinstance(msg, int) else 0) ^ 0x9E3779B97F4A7C15
        for i in range(4096):
            h = ((h * 0xBF58476D1CE4E5B9) + i) & 0xFFFFFFFFFFFFFFFF
            h = ((h << 27) | (h >> 37)) & 0xFFFFFFFFFFFFFFFF
        if isinstance(msg, str) and msg == "__count__":
            return h
        return None


# --- runners ----------------------------------------------------------------


def run_tell(system: ActorSystem, n: int) -> Measurement:
    disp, *_ = best_dispatcher_for("tell")
    ref = system.actor_of(props(Sink, dispatcher=disp), "prof-tell")
    rss0, cpu0 = rss_bytes(), cpu_time_ns()
    t0 = time.perf_counter_ns()
    for i in range(n):
        ref.tell(i)
    total = ref.ask_blocking("__count__", 30.0)
    elapsed_ns = time.perf_counter_ns() - t0
    rss1, cpu1 = rss_bytes(), cpu_time_ns()
    assert total == n, f"expected {n} processed, saw {total}"
    return _make("tell", disp, n, elapsed_ns, (), rss0, rss1, cpu0, cpu1)


def run_ask(system: ActorSystem, n: int) -> Measurement:
    disp, *_ = best_dispatcher_for("ask")
    ref = system.actor_of(props(Echo, dispatcher=disp), "prof-ask")
    samples: List[int] = []
    rss0, cpu0 = rss_bytes(), cpu_time_ns()
    t0 = time.perf_counter_ns()
    for i in range(n):
        s = time.perf_counter_ns()
        ref.ask_blocking(i, 5.0)
        samples.append(time.perf_counter_ns() - s)
    elapsed_ns = time.perf_counter_ns() - t0
    rss1, cpu1 = rss_bytes(), cpu_time_ns()
    return _make("ask", disp, n, elapsed_ns, samples, rss0, rss1, cpu0, cpu1)


def run_fanout(system: ActorSystem, n: int) -> Measurement:
    disp, *_ = best_dispatcher_for("fanout")
    rss0, cpu0 = rss_bytes(), cpu_time_ns()
    t0 = time.perf_counter_ns()
    refs: List[ActorRef] = [
        system.actor_of(props(Sink, dispatcher=disp), f"prof-fan-{i}") for i in range(n)
    ]
    for i, r in enumerate(refs):
        r.tell(i)
    for r in refs:
        assert r.ask_blocking("__count__", 30.0) == 1
    elapsed_ns = time.perf_counter_ns() - t0
    rss1, cpu1 = rss_bytes(), cpu_time_ns()
    return _make("fanout", disp, n, elapsed_ns, (), rss0, rss1, cpu0, cpu1)


def run_cpu(system: ActorSystem, n: int) -> Measurement:
    disp, role, count, quota = best_dispatcher_for("cpu")
    if disp != "python-pinned":
        system.configure_interpreter(role, disp, count, quota)
        ref = system.actor_of(
            props(CpuWorker, dispatcher=disp, interpreter_role=role), "prof-cpu"
        )
        cfg = f"{disp} pool={count}"
    else:
        ref = system.actor_of(props(CpuWorker, dispatcher=disp), "prof-cpu")
        cfg = disp
    rss0, cpu0 = rss_bytes(), cpu_time_ns()
    t0 = time.perf_counter_ns()
    for i in range(n):
        ref.tell(i)
    ref.ask_blocking("__count__", 180.0)
    elapsed_ns = time.perf_counter_ns() - t0
    rss1, cpu1 = rss_bytes(), cpu_time_ns()
    return _make("cpu", cfg, n, elapsed_ns, (), rss0, rss1, cpu0, cpu1)


# --- helpers ----------------------------------------------------------------


def _make(
    scenario: str,
    config: str,
    n: int,
    elapsed_ns: int,
    latencies_ns: Iterable[int],
    rss0: Optional[int],
    rss1: Optional[int],
    cpu0: int,
    cpu1: int,
) -> Measurement:
    throughput = (n * 1e9 / elapsed_ns) if elapsed_ns else 0.0
    lat = sorted(latencies_ns)
    return Measurement(
        runtime="python",
        scenario=scenario,
        config=config,
        messages=n,
        elapsed_ns=elapsed_ns,
        throughput_msgs_per_sec=throughput,
        p50_ns=percentile(lat, 50),
        p95_ns=percentile(lat, 95),
        p99_ns=percentile(lat, 99),
        rss_delta_bytes=(None if (rss0 is None or rss1 is None) else rss1 - rss0),
        peak_rss_bytes=peak_rss_bytes(),
        cpu_delta_ns=max(0, cpu1 - cpu0),
    )

File: crates/rakka-profiler/src/scenarios.rs
//! Scenario implementations. Each returns a [`Measurement`].
//!
//! All scenarios take a pre-built [`ActorSystem`] from the caller so the
//! system-startup cost is attributed to setup rather than the scenario
//! itself.

use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};

use async_trait::async_trait;
use rakka::prelude::*;
use tokio::sync::oneshot;

use crate::metrics::{cpu_delta, cpu_time, peak_rss_bytes, rss_bytes, rss_delta};
use crate::report::{Measurement, Scenario};

/// Fire-and-forget throughput into a null actor.
pub async fn tell(system: &ActorSystem, n: u64) -> anyhow::Result<Measurement> {
    let counter = Arc::new(AtomicUsize::new(0));
    let counter_actor = counter.clone();
    let target = system.actor_of(
        Props::create(move || CountingActor { counter: counter_actor.clone() }),
        "profiler-tell",
    )?;

    let rss_before = rss_bytes();
    let cpu_before = cpu_time();
    let start = Instant::now();
    for i in 0..n {
        target.tell(i);
    }
    while counter.load(Ordering::Acquire) < n as usize {
        tokio::task::yield_now().await;
    }
    let elapsed = start.elapsed();
    let rss_after = rss_bytes();
    let cpu_after = cpu_time();

    Ok(Measurement::from_throughput("rust", Scenario::Tell, "default-dispatcher", n, elapsed)
        .with_memory(rss_delta(rss_before, rss_after), peak_rss_bytes())
        .with_cpu(cpu_delta(cpu_before, cpu_after)))
}

/// Sequential ask latency.
pub async fn ask(system: &ActorSystem, n: u64) -> anyhow::Result<Measurement> {
    let target = system.actor_of(Props::create(|| EchoActor), "profiler-ask")?;

    let rss_before = rss_bytes();
    let cpu_before = cpu_time();
    let start = Instant::now();
    let mut samples: Vec<Duration> = Vec::with_capacity(n as usize);
    for i in 0..n {
        let t0 = Instant::now();
        let _reply: u64 = target
            .ask_with(
                move |reply: oneshot::Sender<u64>| EchoMsg { value: i, reply },
                Duration::from_secs(5),
            )
            .await
            .map_err(|e| anyhow::anyhow!("ask failed: {e:?}"))?;
        samples.push(t0.elapsed());
    }
    let elapsed = start.elapsed();
    let rss_after = rss_bytes();
    let cpu_after = cpu_time();
    samples.sort_unstable();

    Ok(Measurement::from_throughput("rust", Scenario::Ask, "default-dispatcher", n, elapsed)
        .with_latencies(&samples)
        .with_memory(rss_delta(rss_before, rss_after), peak_rss_bytes())
        .with_cpu(cpu_delta(cpu_before, cpu_after)))
}

/// Spawn N actors then hit each with one message; measures creation cost.
pub async fn fanout(system: &ActorSystem, n: u64) -> anyhow::Result<Measurement> {
    let counter = Arc::new(AtomicUsize::new(0));
    let rss_before = rss_bytes();
    let cpu_before = cpu_time();
    let start = Instant::now();

    let mut refs = Vec::with_capacity(n as usize);
    for i in 0..n {
        let c = counter.clone();
        let r = system.actor_of(
            Props::create(move || CountingActor { counter: c.clone() }),
            &format!("profiler-fanout-{i}"),
        )?;
        refs.push(r);
    }
    for (i, r) in refs.iter().enumerate() {
        r.tell(i as u64);
    }
    while counter.load(Ordering::Acquire) < n as usize {
        tokio::task::yield_now().await;
    }
    let elapsed = start.elapsed();
    let rss_after = rss_bytes();
    let cpu_after = cpu_time();

    Ok(Measurement::from_throughput("rust", Scenario::Fanout, "default-dispatcher", n, elapsed)
        .with_memory(rss_delta(rss_before, rss_after), peak_rss_bytes())
        .with_cpu(cpu_delta(cpu_before, cpu_after)))
}

/// CPU-bound handler — each message runs a small hashing loop. Throughput
/// for this scenario reflects per-core compute rather than mailbox speed.
pub async fn cpu(system: &ActorSystem, n: u64) -> anyhow::Result<Measurement> {
    let counter = Arc::new(AtomicUsize::new(0));
    let c = counter.clone();
    let target = system.actor_of(
        Props::create(move || CpuActor { counter: c.clone() }),
        "profiler-cpu",
    )?;

    let rss_before = rss_bytes();
    let cpu_before = cpu_time();
    let start = Instant::now();
    for i in 0..n {
        target.tell(i);
    }
    while counter.load(Ordering::Acquire) < n as usize {
        tokio::task::yield_now().await;
    }
    let elapsed = start.elapsed();
    let rss_after = rss_bytes();
    let cpu_after = cpu_time();

    Ok(Measurement::from_throughput("rust", Scenario::Cpu, "cpu-bound-handler", n, elapsed)
        .with_memory(rss_delta(rss_before, rss_after), peak_rss_bytes())
        .with_cpu(cpu_delta(cpu_before, cpu_after)))
}

// --- actors -----------------------------------------------------------------

struct CountingActor {
    counter: Arc<AtomicUsize>,
}
#[async_trait]
impl Actor for CountingActor {
    type Msg = u64;
    async fn handle(&mut self, _ctx: &mut Context<Self>, _msg: u64) {
        self.counter.fetch_add(1, Ordering::Release);
    }
}

struct EchoMsg {
    value: u64,
    reply: oneshot::Sender<u64>,
}

struct EchoActor;
#[async_trait]
impl Actor for EchoActor {
    type Msg = EchoMsg;
    async fn handle(&mut self, _ctx: &mut Context<Self>, msg: EchoMsg) {
        let _ = msg.reply.send(msg.value);
    }
}

struct CpuActor {
    counter: Arc<AtomicUsize>,
}
#[async_trait]
impl Actor for CpuActor {
    type Msg = u64;
    async fn handle(&mut self, _ctx: &mut Context<Self>, seed: u64) {
        // Deterministic CPU-bound work: 4k iterations of a cheap mixer.
        let mut h: u64 = seed.wrapping_add(0x9E37_79B9_7F4A_7C15);
        for i in 0..4096u64 {
            h ^= h.wrapping_mul(0xBF58_476D_1CE4_E5B9).wrapping_add(i);
            h = h.rotate_left(27);
        }
        std::hint::black_box(h);
        self.counter.fetch_add(1, Ordering::Release);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rakka::prelude::Config;

    #[tokio::test]
    async fn tell_scenario_completes() {
        let sys = ActorSystem::create("t-tell", Config::empty()).await.unwrap();
        let m = tell(&sys, 200).await.unwrap();
        assert_eq!(m.messages, 200);
        assert!(m.throughput_msgs_per_sec > 0.0);
        sys.terminate().await;
    }

    #[tokio::test]
    async fn ask_scenario_has_latencies() {
        let sys = ActorSystem::create("t-ask", Config::empty()).await.unwrap();
        let m = ask(&sys, 50).await.unwrap();
        assert!(m.p50_ns.is_some());
        assert!(m.p95_ns.is_some());
        sys.terminate().await;
    }

    #[tokio::test]
    async fn fanout_spawns_actors() {
        let sys = ActorSystem::create("t-fanout", Config::empty()).await.unwrap();
        let m = fanout(&sys, 20).await.unwrap();
        assert_eq!(m.messages, 20);
        sys.terminate().await;
    }

    #[tokio::test]
    async fn cpu_scenario_runs() {
        let sys = ActorSystem::create("t-cpu", Config::empty()).await.unwrap();
        let m = cpu(&sys, 40).await.unwrap();
        assert!(m.elapsed_ns > 0);
        sys.terminate().await;
    }
}
